staging/sockets/luasockets.c

```diff
@@ -0,0 +1,205 @@
+/* Lua sockets module */
+
+#include <sys/lua.h>
+#include <sys/module.h>
+#include <sys/socketvar.h>
+#include <netinet/in.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+
+#include <stdio.h>
+
+MODULE(MODULE_CLASS_MISC, luasockets, "lua");
+
+#define BUF_SIZE 1024
+
+struct luasocket {
+	struct socket *socket;
+	int fd;
+};
+
+static int
+sockets_socket(lua_State *L)
+{
+	struct luasocket *luasocket;
+	struct socket *so;
+	const char *domain;
+	const char *type;
+	int sfd;
+	int err;
+	int dom = 0;
+	int typ = 0;
+
+	domain = luaL_checkstring(L, 1);
+	type   = luaL_checkstring(L, 2);
+
+	//XXX not about using strings to name domains (types, etc)
+	if (strcmp(domain, "inet4") == 0)
+		dom = AF_INET;
+	else if (strcmp(domain, "inet6"))
+		dom = AF_INET6;
+	else if (strcmp(domain, "local"))
+		dom = AF_LOCAL;
+	else
+		luaL_error(L, "'%s' domain not supported\n", domain);
+
+	if (strcmp(type, "dgram") == 0 || strcmp(type, "datagram") == 0)
+		typ = SOCK_DGRAM;
+	else if (strcmp(type, "stream"))
+		typ = SOCK_STREAM;
+	else
+		luaL_error(L, "'%s' type not supported\n", type);
+
+	if ((err = fsocreate(dom, &so, typ, 0, &sfd)))
+		luaL_error(L, "could not create socket (error %d)", err);
+
+	luasocket = (struct luasocket*)
+		lua_newuserdata(L, sizeof(struct luasocket));
+	luasocket->socket = so;
+	luasocket->fd = sfd;
+
+	return 1;
+}
+
+static int
+sockets_close(lua_State *L)
+{
+	struct luasocket *luasocket;
+	int err;
+
+	if ((luasocket = (struct luasocket *) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid Lua socket\n");
+	if ((err = soclose(luasocket->socket)) != 0)
+		luaL_error(L, "could not close socket (error %d)\n", err);
+	lua_pushboolean(L, 1);
+	return 1;
+}
+
+typedef int BindOrConnect(struct socket*, struct sockaddr*, struct lwp*);
```

Comment:
Perhaps it'd be the case if we had more than 2 functions sharing the same interface? In this case I think it just makes it obscure -- it's easier to grasp what BindOrConnect means.